network/src/compress.rs

```diff
@@ -116,3 +117,142 @@ pub fn compress(src: Bytes) -> Bytes {
+                                "succeeded",
+                                "compressed ratio",
+                            ])
+                            .observe(res.len() as f64 / data.len() as f64);
+                    }
+                    // compressed data is larger than or equal to uncompressed data
+                    if res.len() >= data.len() {
+                        process(&data, UNCOMPRESS_FLAG, dst)?;
+                    } else {
+                        process(&res, COMPRESS_FLAG, dst)?;
+                    }
+                }
+                Err(e) => {
+                    debug!(
+                        "protocol {} message snappy compress error: {}",
+                        self.protocol_id, e
+                    );
+                    if let Some(metrics) = ckb_metrics::handle() {
+                        metrics
+                            .ckb_network_compress
+                            .with_label_values(&[
+                                self.protocol_id.to_string().as_str(),
+                                "failed",
+                                "compressed ratio",
+                            ])
+                            .observe(1.0);
+                    }
+                    process(&data, UNCOMPRESS_FLAG, dst)?;
+                }
+            }
+        } else {
+            if let Some(metrics) = ckb_metrics::handle() {
+                metrics
+                    .ckb_network_not_compress_count
+                    .with_label_values(&[self.protocol_id.to_string().as_str()])
+                    .inc();
+            }
+            process(&data, UNCOMPRESS_FLAG, dst)?;
+        }
+        Ok(())
+    }
+}
+
+impl tokio_util::codec::Decoder for LengthDelimitedCodecWithCompress {
+    type Item = BytesMut;
+    type Error = io::Error;
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<BytesMut>, io::Error> {
+        if src.is_empty() {
+            return Ok(None);
+        }
+        match self.length_delimited.decode(src)? {
+            Some(mut data) => {
+                if (data[0] & COMPRESS_FLAG) != 0 {
+                    match decompress_len(&data[1..]) {
+                        Ok(decompressed_bytes_len) => {
+                            if decompressed_bytes_len > MAX_UNCOMPRESSED_LEN {
+                                debug!(
+                                    "The limit for uncompressed bytes len is exceeded. limit: {}, len: {}",
+                                    MAX_UNCOMPRESSED_LEN, decompressed_bytes_len
+                                );
+                                return Err(io::ErrorKind::InvalidData.into());
+                            }
+                            let mut buf = vec![0; decompressed_bytes_len];
+                            match SnapDecoder::new().decompress(&data[1..], &mut buf) {
+                                Ok(_) => Ok(Some(BytesMut::from(Bytes::from(buf)))),
```

Comment:
here is a data copy when convert buf to Bytes, can we use decompress_vec to write the data to BytesMut directly?